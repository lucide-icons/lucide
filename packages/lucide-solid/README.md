# Lucide Solid

Implementation of the lucide icon library for solid applications.

> What is lucide? Read it [here](https://github.com/lucide-icons/lucide#what-is-lucide).

## Installation

```sh
yarn add lucide-solid
```

or

```sh
npm install lucide-solid
```

## How to use

It's build with ESmodules so it's completely tree-shakable.
Each icon can be imported as a solid component.

### Example

You can pass additional props to adjust the icon.

```js
import { Camera } from 'lucide-solid';
// Returns SolidComponent

// Usage
const App = () => {
  return <Camera color="red" size={48} />;
};

export default App;
```

### Props

| name          | type     | default      |
| ------------- | -------- | ------------ |
| `size`        | _Number_ | 24           |
| `color`       | _String_ | currentColor |
| `strokeWidth` | _Number_ | 2            |

### Custom props / svg attributes

You can also pass custom props that will be added in the as attributes. With that you can modify the icons look by passing svg attributes.

```js
// Usage
const App = () => {
  return <Camera fill="red" stroke-linejoin="bevel" />;
};
```

### One generic icon component

It is possible to create one generic icon component to load icons.

> :warning: Example below importing all EsModules, caution using this example, not recommended when you using bundlers, your application build size will grow strongly.

#### Icon Component Example

```tsx
import * as icons from 'lucide-solid';
import type { LucideProps } from 'lucide-solid';
import { splitProps } from 'solid-js';
import { Dynamic } from 'solid-js/web';

const Icon = (props: { name: keyof typeof icons } & LucideProps) => {
  const [local, others] = splitProps(props, ["name"]);

  return <Dynamic component={icons[local.name]} {...others} />
};

export default Icon;
```
