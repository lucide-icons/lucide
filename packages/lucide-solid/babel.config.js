const mainConfig = require('../../babel.config');

module.exports = {
  presets: ['solid'],
  env: mainConfig.env,
};
