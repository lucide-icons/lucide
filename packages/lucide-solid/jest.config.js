module.exports = {
  verbose: true,
  // preset: 'solid-jest/preset/browser',
  roots: ['<rootDir>/src/', '<rootDir>/tests/'],
  moduleFileExtensions: ['js'],
  transformIgnorePatterns: [`/node_modules`],
  transform: {
    '^.+\\.js$': 'babel-jest',
  },
};
