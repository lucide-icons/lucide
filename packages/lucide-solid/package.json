{
  "name": "lucide-solid",
  "description": "A Lucide icon library package for Solid applications",
  "version": "0.88.0-beta.3",
  "license": "ISC",
  "homepage": "https://lucide.dev",
  "bugs": "https://github.com/lucide-icons/lucide/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/lucide-icons/lucide.git",
    "directory": "packages/lucide-solid"
  },
  "author": "Eric Fennis",
  "amdName": "lucide-solid",
  "main": "dist/cjs/lucide-solid.js",
  "main:umd": "dist/umd/lucide-solid.js",
  "module": "dist/esm/lucide-solid.js",
  "unpkg": "dist/umd/lucide-solid.min.js",
  "typings": "dist/lucide-solid.d.ts",
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "scripts": {
    "build": "pnpm clean && pnpm copy:license && pnpm build:icons && pnpm build:bundles && pnpm build:types",
    "copy:license": "cp ../../LICENSE ./LICENSE",
    "clean": "rm -rf dist && rm -rf stats && rm -rf ./src/icons/*.js",
    "build:icons": "node ../../scripts/buildIcons.mjs --output=./src --templateSrc=./scripts/exportTemplate.mjs --renderUniqueKey --iconFileExtension=.ts --exportFileName=index.ts",
    "build:types": "node ./scripts/buildTypes.mjs",
    "build:bundles": "rollup -c ./rollup.config.mjs",
    "test": "vitest run",
    "version": "pnpm version --git-tag-version=false"
  },
  "devDependencies": {
    "@lucide/rollup-plugins": "workspace:*",
    "@testing-library/jest-dom": "^5.16.5",
    "babel-preset-solid": "^1.5.4",
    "jsdom": "^20.0.0",
    "rollup": "^3.5.1",
    "solid-js": "^1.4.7",
    "solid-testing-library": "^0.3.0",
    "vite": "^3.2.4",
    "vite-plugin-solid": "^2.3.0",
    "vitest": "^0.23.2"
  },
  "peerDependencies": {
    "solid-js": "^1.4.7"
  }
}
