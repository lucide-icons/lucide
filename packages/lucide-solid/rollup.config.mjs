import plugins from '@lucide/rollup-plugins';
import pkg from './package.json' assert { type: 'json' };

const packageName = 'LucideSolid';
const outputFileName = 'lucide-solid';
const outputDir = 'dist';
const inputs = [`src/lucide-solid.ts`];
const bundles = [
  {
    format: 'umd',
    inputs,
    outputDir,
    minify: true,
  },
  {
    format: 'umd',
    inputs,
    outputDir,
  },
  {
    format: 'cjs',
    inputs,
    outputDir,
  },
  {
    format: 'es',
    inputs,
    outputDir,
  },
  {
    format: 'esm',
    inputs,
    outputDir,
    preserveModules: true,
  },
];

const configs = bundles
  .map(({ inputs, outputDir, format, minify, preserveModules }) =>
    inputs.map(input => ({
      input,
      plugins: plugins(pkg, minify),
      external: ['solid-js', 'solid-js/h'],
      output: {
        name: packageName,
        ...(preserveModules
          ? {
              dir: `${outputDir}/${format}`,
            }
          : {
              file: `${outputDir}/${format}/${outputFileName}${minify ? '.min' : ''}.js`,
            }),
        preserveModules,
        format,
        sourcemap: true,
        globals: {
          'solid-js': 'solid-js',
          'solid-js/h': 'solid-js/h',
        },
      },
    })),
  )
  .flat();

export default configs;
