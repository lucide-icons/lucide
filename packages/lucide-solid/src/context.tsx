import { createContext, splitProps, type JSXElement } from 'solid-js';

export const LucideContext = createContext<{
  size?: number;
  fill?: string;
  color?: string;
  strokeWidth?: number;
  absoluteStrokeWidth?: boolean;
}>({
  size: 24,
  fill: 'none',
  color: 'currentColor',
  strokeWidth: 2,
  absoluteStrokeWidth: false,
});

interface LucideProviderProps {
  children: JSXElement;
  size?: number;
  fill?: string;
  color?: string;
  strokeWidth?: number;
  absoluteStrokeWidth?: boolean;
}

export function LucideProvider(props: LucideProviderProps) {
  const [value, rest] = splitProps(props, ['size', 'color', 'strokeWidth', 'absoluteStrokeWidth']);

  return <LucideContext.Provider value={value}>{rest.children}</LucideContext.Provider>;
}
