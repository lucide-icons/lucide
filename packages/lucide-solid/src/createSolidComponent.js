import h from 'solid-js/h';
import { mergeProps } from 'solid-js';
import defaultAttributes from './defaultAttributes';

/**
 * Converts string to KebabCase
 * Copied from scripts/helper. If anyone knows how to properly import it here
 * then please fix it.
 *
 * @param {string} string
 * @returns {string} A kebabized string
 */
export const toKebabCase = string => string.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();

export default (iconName, iconNode) => {
  const Component = ({
    color = 'currentColor',
    size = 24,
    strokeWidth = 2,
    children = [],
    ...rest
  }) =>
    h(
      'svg',
      mergeProps(
        defaultAttributes,
        {
          width: size,
          height: size,
          stroke: color,
          'stroke-width': strokeWidth,
          classList: {
            lucide: true,
            [`lucide-${toKebabCase(iconName)}`]: true,
          },
        },
        rest,
      ),
      [...iconNode.map(([tag, attrs]) => h(tag, attrs)), children],
    );

  Component.displayName = `${iconName}`;

  return Component;
};
