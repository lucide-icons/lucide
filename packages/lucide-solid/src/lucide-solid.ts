/**
 * lucide-solid {{version}} - ISC
 */

export * from './icons';
export * as icons from './icons';
export * from './aliases';
