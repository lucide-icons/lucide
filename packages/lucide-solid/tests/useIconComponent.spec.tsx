import { describe, it, expect } from 'vitest';
import { useIconComponent } from '../src/lucide-solid';
import { airVent, coffee } from './testIconNodes';
import { render } from '@solidjs/testing-library';

describe('Using iconNode component generator', () => {
  it('should create a component from an iconNode', () => {
    const { AirVent } = useIconComponent({ airVent });

    const { container } = render(() => <AirVent />);

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toBeDefined();
  });

  it('should create multiple components', () => {
    const { AirVent, Coffee } = useIconComponent({ airVent, coffee });

    const { container } = render(() => (
      <>
        <AirVent />
        <Coffee color="blue" />
      </>
    ));

    expect(container).toMatchSnapshot();
    expect(container).toBeDefined();
  });
});
