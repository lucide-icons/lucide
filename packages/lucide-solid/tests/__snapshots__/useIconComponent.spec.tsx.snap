// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Using iconNode component generator > should create a component from an iconNode 1`] = `
<svg
  class="lucide lucide-icon lucide-air-vent"
  fill="none"
  height="24"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M6 12H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v5a2 2 0 0 1-2 2h-2"
    key="larmp2"
  />
  <path
    d="M6 8h12"
    key="6g4wlu"
  />
  <path
    d="M18.3 17.7a2.5 2.5 0 0 1-3.16 3.83 2.53 2.53 0 0 1-1.14-2V12"
    key="1bo8pg"
  />
  <path
    d="M6.6 15.6A2 2 0 1 0 10 17v-5"
    key="t9h90c"
  />
</svg>
`;

exports[`Using iconNode component generator > should create multiple components 1`] = `
<div>
  <svg
    class="lucide lucide-icon lucide-air-vent"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M6 12H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v5a2 2 0 0 1-2 2h-2"
      key="larmp2"
    />
    <path
      d="M6 8h12"
      key="6g4wlu"
    />
    <path
      d="M18.3 17.7a2.5 2.5 0 0 1-3.16 3.83 2.53 2.53 0 0 1-1.14-2V12"
      key="1bo8pg"
    />
    <path
      d="M6.6 15.6A2 2 0 1 0 10 17v-5"
      key="t9h90c"
    />
  </svg>
  <svg
    class="lucide lucide-icon lucide-coffee"
    fill="none"
    height="24"
    stroke="blue"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M17 8h1a4 4 0 1 1 0 8h-1"
      key="jx4kbh"
    />
    <path
      d="M3 8h14v9a4 4 0 0 1-4 4H7a4 4 0 0 1-4-4Z"
      key="1bxrl0"
    />
    <line
      key="1cr9l3"
      x1="6"
      x2="6"
      y1="2"
      y2="4"
    />
    <line
      key="170wym"
      x1="10"
      x2="10"
      y1="2"
      y2="4"
    />
    <line
      key="1c5f70"
      x1="14"
      x2="14"
      y1="2"
      y2="4"
    />
  </svg>
</div>
`;
